npm/pkg/transport/transport.go

```diff
@@ -0,0 +1,112 @@
+package transport
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/Azure/azure-container-networking/npm/pkg/transport/pb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+	"google.golang.org/grpc/stats"
+	"k8s.io/klog/v2"
+)
+
+type Manager struct {
+	// Server is the gRPC server
+	Server pb.DataplaneEventsServer
+
+	// Watchdog is the watchdog for the gRPC server that implements the
+	// gRPC stats handler interface
+	Watchdog stats.Handler
+
+	// Registrations is a map of dataplane pod address to their associate connection stream
+	Registrations map[string]clientStreamConnection
+
+	// port is the port the manager is listening on
+	port int
+
+	// regCh is the registration channel
+	regCh chan clientStreamConnection
+
+	// deregCh is the deregistration channel
+	deregCh chan string
+
+	// errCh is the error channel
+	errCh chan error
+}
+
+// New creates a new transport manager
+func NewManager(port int) *Manager {
+	// Create a registration channel
+	regCh := make(chan clientStreamConnection, grpcMaxConcurrentStreams)
+
+	// Create a deregistration channel
+	deregCh := make(chan string)
+
+	return &Manager{
+		// Broadcaster: gbc.NewBroadcaster(concurrentInputRegistrations),
+		Server:        NewServer(regCh),
+		Watchdog:      NewWatchdog(deregCh),
+		Registrations: make(map[string]clientStreamConnection),
+		port:          port,
+		errCh:         make(chan error),
+		deregCh:       deregCh,
+		regCh:         regCh,
+	}
+}
+
+func (m *Manager) Start(ctx context.Context) {
+	klog.Info("Starting transport manager")
+	m.start(ctx)
+}
+
+func (m *Manager) start(ctx context.Context) {
+	go m.handle()
+
+	for {
+		select {
+		case client := <-m.regCh:
+			klog.Infof("Registering remote client %s", client)
+			m.Registrations[client.String()] = client
+		case client := <-m.deregCh:
```

Comment:
there can some potential race conditions here ?  what if a connection got lost and created for the same pod in immediate succession, and lets say deregCh is overloaded and regCh is free, then there is a high chance that we would delete a legitimate connection right ? how do we manage those kinds of race conditions ?

We might need to introduce an aspect of timestamp ?

Reply:
I think a ping-ping health check similar to Felix and typha is a must to avoid such issues. We need some keepalives and checks in place. Thoughts?

Reply:
Yea I think timestamps makes sense.